pkg/api/customalerting/v1alpha1/types.go

```diff
@@ -0,0 +1,389 @@
+/*
+Copyright 2020 KubeSphere Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/emicklei/go-restful"
+	"github.com/pkg/errors"
+	prommodel "github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/promql/parser"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+const (
+	RuleLevelCluster   RuleLevel = "cluster"
+	RuleLevelNamespace RuleLevel = "namespace"
+)
+
+var (
+	ErrThanosRulerNotEnabled     = errors.New("The request operation to custom alerting rule could not be done because thanos ruler is not enabled")
+	ErrAlertingRuleNotFound      = errors.New("The alerting rule was not found")
+	ErrAlertingRuleAlreadyExists = errors.New("The alerting rule already exists")
+
+	ruleLabelNameMatcher = regexp.MustCompile(`[a-zA-Z_][a-zA-Z0-9_]*`)
+)
+
+type RuleLevel string
+
+type AlertingRuleQualifier struct {
+	Id     string    `json:"id,omitempty" description:"rule id is mainly for the builtin rules"`
+	Name   string    `json:"name,omitempty" description:"rule name which should be uniq for custom rules in the specified namespace"`
+	Level  RuleLevel `json:"level,omitempty" description:"rule level is only for the custom rules and its value is one of cluster, namespace"`
+	Custom bool      `json:"custom" description:"whether to be a custom rule. The builtin rules are not custom and can only be viewed"`
+}
+
+type AlertingRuleProps struct {
```

Comment:
As discussed we can adjust AlertingRuleProps PostableAlertingRule GettableAlertingRule as below:
- AlertingRuleProps => AlertingRule
- Move common properties like Alias and Description to AlertingRule
- Move name and id in AlertingRuleQualifier to AlertingRule
- PostableAlertingRule could be replaced by AlertingRule
- Custom and Level could be removed from AlertingRuleQualifier
- Change Rule *AlertingRuleQualifier in Alerts to RuleId and RuleName